pkg/controller/pxc/users.go

```diff
@@ -38,159 +35,355 @@ type ReconcileUsersResult struct {
-		if cr.CompareVersionWith("1.9.0") >= 0 {
-			err = r.manageReplicationUser(cr, &sysUsersSecretObj, &internalSysSecretObj)
-			if err != nil {
-				return nil, errors.Wrap(err, "manage replication user")
+		case users.UserMonitor:
+			if err := r.handleMonitorUser(cr, secrets, internalSecrets, res); err != nil {
+				return res, err
+			}
+		case users.UserClustercheck:
+			if err := r.handleClustercheckUser(cr, secrets, internalSecrets, res); err != nil {
+				return res, err
+			}
+		case users.UserXtrabackup:
+			if err := r.handleXtrabackupUser(cr, secrets, internalSecrets, res); err != nil {
+				return res, err
+			}
+		case users.UserReplication:
+			if err := r.handleReplicationUser(cr, secrets, internalSecrets, res); err != nil {
+				return res, err
+			}
+		case users.UserProxyAdmin:
+			if err := r.handleProxyadminUser(cr, secrets, internalSecrets, res); err != nil {
+				return res, err
+			}
+		case users.UserPMMServer, users.UserPMMServerKey:
+			if err := r.handlePMMUser(cr, secrets, internalSecrets, res); err != nil {
+				return res, err
 			}
 		}
+	}
 
-		if cr.CompareVersionWith("1.10.0") >= 0 {
-			mysqlVersion := cr.Status.PXC.Version
-			if mysqlVersion == "" {
-				mysqlVersion, err = r.mysqlVersion(cr, statefulset.NewNode(cr))
-				if err != nil && !errors.Is(err, versionNotReadyErr) {
-					return nil, errors.Wrap(err, "retrieving pxc version")
-				}
-			}
+	return res, nil
+}
 
-			if mysqlVersion != "" {
-				ver, err := version.NewVersion(mysqlVersion)
-				if err != nil {
-					return nil, errors.Wrap(err, "invalid pxc version")
-				}
+func (r *ReconcilePerconaXtraDBCluster) handleRootUser(cr *api.PerconaXtraDBCluster, secrets, internalSecrets *corev1.Secret, actions *userUpdateActions) error {
+	logger := r.logger(cr.Name, cr.Namespace)
 
-				if !ver.LessThan(privSystemUserAddedIn) {
-					if err := r.grantSystemUserPrivilege(cr, &internalSysSecretObj); err != nil {
-						return nil, errors.Wrap(err, "grant system privilege")
-					}
-				}
-			}
-		}
+	user := &users.SysUser{
+		Name:  users.UserRoot,
+		Pass:  string(secrets.Data[users.UserRoot]),
+		Hosts: []string{"localhost", "%"},
 	}
 
-	if cr.Status.Status != api.AppStateReady {
-		return nil, nil
+	if bytes.Equal(secrets.Data[user.Name], internalSecrets.Data[user.Name]) {
+		return nil
+	}
+
+	logger.Info(fmt.Sprintf("User %s: password changed, updating user", user.Name))
+
+	// update pass
+	err := r.updateUserPass(cr, secrets, internalSecrets, user)
+	if err != nil {
+		return errors.Wrap(err, "update root users pass")
 	}
+	logger.Info(fmt.Sprintf("User %s: password updated", user.Name))
 
-	newSysData, err := json.Marshal(sysUsersSecretObj.Data)
+	// syncProxyUser
+	err = r.syncPXCUsersWithProxySQL(cr)
 	if err != nil {
-		return nil, errors.Wrap(err, "marshal sys secret data")
+		return errors.Wrap(err, "sync users")
 	}
-	newSecretDataHash := sha256Hash(newSysData)
+	logger.Info(fmt.Sprintf("User %s: synced PXC users with ProxySQL", user.Name))
```

Comment:
should we print this even proxysql is not enabled?

Reply:
Well, I guess if it is not enabled, we shouldn't sync anything, not just print this.